lib/ansible/inventory/manager.py

```diff
@@ -260,6 +263,10 @@ def parse_source(self, source, cache=False):
                     plugin_wants = False
 
                 if plugin_wants:
+                    # No need to deepcopy inventory unless it has changed
+                    if inventory_copy is None:
+                        inventory_copy = deepcopy(self._inventory)
```

Comment:
As a note to myself and @chrismeyersfsu, this is a motivator for me want to check running this with multi-10s of thousands of hosts.

Reply:
Yeah, this line was the cause of decoupling the objects because it was not scaling well. I generated a mock inventory directory (20 files, 10,000 hosts a file, failing from invalid syntax at the very end of every other file to test correct inventory restoration) and added a cgroup profiler to the inventory manager. I'm still looking for optimizations, so let me know your findings/ideas when you test.